client/src/main/scala/org/http4s/client/Client.scala

```diff
@@ -319,6 +319,15 @@ object Client {
     Client(disposableService(service), F.delay(isShutdown.set(true)))
   }
 
+  private def addHostHeaderIfUriIsAbsolute[F[_]](req: Request[F]): Request[F] = {
+    val maybeHost = for {
+      host <- req.uri.host
+      if req.headers.get(Host).isEmpty
```

Comment:
`_ <- Alternative[Option].guard(req.headers.get(Host).isEmpty)`

Reply:
@ChristopherDavenport I don't understand the difference here. Can you explain why `Alternative[Option].guard` is preferred?

Reply:
One is implemented in terms of `flatMap` the other in terms of `withFilter`. I just see this as utilizing non-functional language features which has the possibility for issues in the future.

Reply:
No comment on cats fp syntax (I don't care for it but I don't hate it), however, instead of the withfilter guard, you can always just

```scala
req.uri.host match {
  case Some(host) if req.headers.get(Host).isEmpty => 
     req.withHeaders(req.headers.put(host))
  case _ => req
}
```
That's microscopically more efficient (no fold call + no withFilter)

Reply:
What is the functional objection to `withFilter` on a type that has a reasonable empty?  I agree that @jmcardon's version saves a few nanoseconds, and I agree that `withFilter` on `Future` is regrettable, but I don't have a problem with what @bpholt wrote.